app/dot/cpp/repository/services/RepositoryService.java

```diff
@@ -17,59 +19,82 @@
+      createCollections(database, entities, withHistory);
     }
   }
 
   private void createCollections(
-      List<Class<? extends BaseEntity>> entities, MongoDatabase database) {
+      MongoDatabase database, List<Class<? extends BaseEntity>> entities, boolean withHistory) {
     final var schemaGeneratorConfig = getSchemaGeneratorConfig();
     entities.forEach(
         entity -> {
           var schema = getSchema(entity, schemaGeneratorConfig);
           var validationOptions =
               new ValidationOptions().validator(Filters.jsonSchema(Document.parse(schema)));
 
-          if (isCollectionInDatabase(entity.getSimpleName(), database)) {
-            logger.debug("already exists");
-
-            database.runCommand(
-                new Document("collMod", entity.getSimpleName())
-                    .append("validator", Filters.jsonSchema(Document.parse(schema)))
-                    .append("validationLevel", "strict"));
-            // default values are added by setting the variable in its respective class with an
-            // initial value
-          } else {
-            database.createCollection(
-                entity.getSimpleName(),
-                new CreateCollectionOptions().validationOptions(validationOptions));
+          createCollection(database, entity.getSimpleName(), schema, validationOptions);
+
+          if (withHistory) {
+            createCollection(
+                database, entity.getSimpleName() + "_history", schema, validationOptions);
+
+            // todo uncomment after history is implemented throughout the application
+            // createIndex(database, entity.getSimpleName(), "trackingId", true);
+            createIndex(database, entity.getSimpleName() + "_history", "trackingId", false);
           }
 
           logger.debug("{}", schema);
         });
   }
 
+  private void createIndex(
+      MongoDatabase database, String entityName, String fieldName, Boolean unique) {
+    final var collection = database.getCollection(entityName);
+
+    for (final var index : collection.listIndexes()) {
+      if (index.getString("name").equals(fieldName + "_idx")) {
+        collection.dropIndex(fieldName + "_idx");
+        break;
+      }
+    }
+
+    collection.createIndex(
+        Indexes.ascending(fieldName), new IndexOptions().name(fieldName + "_idx").unique(unique));
+  }
+
+  private void createCollection(
+      MongoDatabase database,
+      String entityName,
+      String schema,
+      ValidationOptions validationOptions) {
+    if (isCollectionInDatabase(entityName, database)) {
+      logger.debug("already exists");
+
+      database.runCommand(
+          new Document("collMod", entityName)
+              .append("validator", Filters.jsonSchema(Document.parse(schema)))
+              .append("validationLevel", "strict"));
+      // default values are set by hard coding an initial value for those variables
```

Comment:
?

Reply:
This comment was actually here before, I just changed it up a little. It's just a note and it could be useful when we create a documentation.